hydroflow_lang/src/graph/mod.rs

```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
+                        || (node_union.same_set(x_src, dst) && node_union.same_set(x_dst, src))
+                })
+            {
+                continue;
+            }
 
-                // Set `src` or `dst` color if `None` based on the other (if possible):
-                // Pull -> Pull
-                // Push -> Push
-                // Pull -> [Comp] -> Push
-                // Push -> [Hoff] -> Pull
-                match (node_color[src], node_color[dst]) {
-                    (Some(_), Some(_)) => (),
-                    (None, None) => (),
-                    (None, Some(dst_color)) => {
-                        node_color[src] = Some(match dst_color {
-                            Color::Comp => Color::Pull,
-                            Color::Hoff => Color::Push,
-                            pull_or_push => pull_or_push,
-                        });
-                        updated = true;
-                    }
-                    (Some(src_color), None) => {
-                        node_color[dst] = Some(match src_color {
-                            Color::Comp => Color::Push,
-                            Color::Hoff => Color::Pull,
-                            pull_or_push => pull_or_push,
-                        });
-                        updated = true;
-                    }
+            // Set `src` or `dst` color if `None` based on the other (if possible):
+            // Pull -> Pull
+            // Push -> Push
+            // Pull -> [Comp] -> Push
+            // Push -> [Hoff] -> Pull
+            match (node_color[src], node_color[dst]) {
+                (Some(_), Some(_)) => (),
+                (None, None) => (),
+                (None, Some(dst_color)) => {
+                    node_color[src] = Some(match dst_color {
+                        Color::Comp => Color::Pull,
+                        Color::Hoff => Color::Push,
+                        pull_or_push => pull_or_push,
+                    });
+                    updated = true;
                 }
+                (Some(src_color), None) => {
+                    node_color[dst] = Some(match src_color {
+                        Color::Comp => Color::Push,
+                        Color::Hoff => Color::Pull,
+                        pull_or_push => pull_or_push,
+                    });
+                    updated = true;
+                }
+            }
 
-                // If SRC and DST can be in the same subgraph.
-                let can_connect = match (node_color[src], node_color[dst]) {
-                    (Some(Color::Pull), Some(Color::Pull)) => true,
-                    (Some(Color::Pull), Some(Color::Comp)) => true,
-                    (Some(Color::Pull), Some(Color::Push)) => true,
+            // If SRC and DST can be in the same subgraph.
+            let can_connect = match (node_color[src], node_color[dst]) {
+                (Some(Color::Pull), Some(Color::Pull)) => true,
```

Comment:
With some work I can read that there are 5x5 possibilities and you've handled all of them. I wonder if there's a more readable/checkable way to do that? Possibly encode it as an explicit table we can inspect as text?

Reply:
Also factor it out of this routine, which is long.